Binary Logistic Regression and the Challenges of Messy Data: A Case of Disappointing Results


--What Data to Keep and What to Delete 

--Process of Cleaning Data

-- Model to Predict Fatal/Non-Fatal Shark Attacks by Geographical Region and Other Variables (Attempt)

-- Original Dataset: https://www.kaggle.com/datasets/felipeesc/shark-attack-dataset

In [692]:
import pandas as pd
url = 'https://raw.githubusercontent.com/A-Bin1/Statistical-Samples/main/shark-attack-data.csv'
sharkdf = pd.read_csv(url, encoding = 'unicode_escape', engine ='python')

In [693]:
sharkdf.head()

Date    Year        Type    Country             Area   
0  25-Jun-18  2018.0     Boating        USA       California  \
1  18-Jun-18  2018.0  Unprovoked        USA          Georgia   
2   9-Jun-18  2018.0     Invalid        USA           Hawaii   
3   8-Jun-18  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4   4-Jun-18  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age   
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57  \
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)   
0  No injury to occupant, outrigger canoe and pad...           N  \
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species  Unnamed: 14 Unnamed: 15  
0          18h00      White shark         NaN         NaN  
1  14h00  -15h00              NaN         NaN         NaN  
2          07h45              NaN         NaN         NaN  
3            NaN        2 m shark         NaN         NaN  
4            NaN  Tiger shark, 3m         NaN         NaN

Clean the dataset: remove unnecessary columns, reformat columns, etc.

In [694]:
del sharkdf['Unnamed: 14']
del sharkdf['Unnamed: 15']

In [695]:
sharkdf.head()

Date    Year        Type    Country             Area   
0  25-Jun-18  2018.0     Boating        USA       California  \
1  18-Jun-18  2018.0  Unprovoked        USA          Georgia   
2   9-Jun-18  2018.0     Invalid        USA           Hawaii   
3   8-Jun-18  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4   4-Jun-18  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age   
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57  \
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)   
0  No injury to occupant, outrigger canoe and pad...           N  \
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species   
0          18h00      White shark  
1  14h00  -15h00              NaN  
2          07h45              NaN  
3            NaN        2 m shark  
4            NaN  Tiger shark, 3m

In [611]:
sharkdf.isnull().sum() > 1000

Date           False
Year           False
Type           False
Country        False
Area           False
Location       False
Activity       False
Name           False
Sex            False
Age             True
Injury         False
Fatal (Y/N)    False
Time            True
Species         True
dtype: bool

In [612]:
del sharkdf['Year']
del sharkdf['Time']
del sharkdf['Species ']
del sharkdf['Age']
del sharkdf['Name']

In [613]:
sharkdf.groupby(['Fatal (Y/N)'], as_index=False).count()

Fatal (Y/N)  Date  Type  Country  Area  Location  Activity  Sex   Injury
0           N     7     7        7     7         7         5     7       7
1        2017     1     1        1     1         1         1     1       1
2           M     1     1        1     1         1         1     1       1
3           N  4293  4292     4274  4087      4018      4018  3945    4293
4          N      1     1        1     1         1         0     0       1
5     UNKNOWN    71    71       69    64        60        36    40      51
6           Y  1388  1386     1364  1198      1191      1282  1281    1387
7           y     1     1        1     0         1         1     1       1

In [614]:
#regroup the Fatal Column
sharkdf.loc[sharkdf['Fatal (Y/N)'] == 'M']

Date      Type           Country      Area      Location   
285  18-Apr-16  Provoked  FRENCH POLYNESIA  Tuamotos  Makemo Atoll  \

         Activity Sex                                              Injury   
285  Spearfishing    M  Laceration to knee by speared shark PROVOKED I...  \

    Fatal (Y/N)  
285           M

In [615]:
sharkdf['Injury'].iloc[285]  #row 285 did not die.

'Laceration to knee by speared shark PROVOKED INCIDENT'

In [616]:
sharkdf.loc[sharkdf['Fatal (Y/N)'] == 'y'] #row 6117

Date        Type Country Area         Location Activity Sex    
6117  October 1815  Unprovoked  TOBAGO  NaN  Scarborough Bay  Bathing    M  \

     Injury Fatal (Y/N)  
6117  FATAL           y

In [617]:
sharkdf.loc[sharkdf['Fatal (Y/N)'] == '2017']  #row 786 did not die

Date      Type Country      Area  Location   
786  10-Jun-12  Provoked   ITALY  Sardinia  Muravera  \

                                            Activity Sex    
786  Attempting to rescue an injured & beached shark    M  \

                                       Injury Fatal (Y/N)  
786  Lower left leg injured PROVOKED ACCIDENT        2017

In [618]:
sharkdf['Fatal (Y/N)'] = sharkdf['Fatal (Y/N)'].replace(['2017','M'],'N')
sharkdf['Fatal (Y/N)'] = sharkdf['Fatal (Y/N)'].replace(['y'],'Y')

In [623]:
sharkdf.head()

Date        Type    Country             Area   
0  25-Jun-18     Boating        USA       California  \
1  18-Jun-18  Unprovoked        USA          Georgia   
2   9-Jun-18     Invalid        USA           Hawaii   
3   8-Jun-18  Unprovoked  AUSTRALIA  New South Wales   
4   4-Jun-18    Provoked     MEXICO           Colima   

                         Location     Activity Sex    
0     Oceanside, San Diego County     Paddling    F  \
1  St. Simon Island, Glynn County     Standing    F   
2                    Habush, Oahu      Surfing    M   
3              Arrawarra Headland      Surfing    M   
4                        La Ticla  Free diving    M   

                                              Injury Fatal (Y/N)  
0  No injury to occupant, outrigger canoe and pad...           N  
1                         Minor injury to left thigh           N  
2       Injury to left lower leg from surfboard skeg           N  
3                          Minor injury to lower leg           N  
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N

In [620]:
print(sharkdf['Fatal (Y/N)'].iloc[786], sharkdf['Fatal (Y/N)'].iloc[6117], sharkdf['Fatal (Y/N)'].iloc[285] )

N Y N


In [624]:
sharkdf = sharkdf.drop(sharkdf['Injury'].loc[sharkdf['Fatal (Y/N)'] == 'UNKNOWN'].index) #deleting unknown for organizational sake

In [625]:
#verify the notes of Injury match Fatal (Y/N):

sharkdf.dropna()[sharkdf.dropna()['Injury'].str.contains('FATAL') & sharkdf.dropna()['Fatal (Y/N)'].str.contains('N') ]

Date        Type Country     Area   Location Activity Sex    
6182  1990 or 1991  Unprovoked   KENYA  Mombasa  Kilindini   Diving    M  \

                                              Injury Fatal (Y/N)  
6182  Conway's leg was bitten Higgs injury was FATAL           N

In [630]:
sharkdf = sharkdf.replace(sharkdf.iloc[6182, sharkdf.columns.get_loc('Fatal (Y/N)')],'Y')

In [631]:
sharkdf.groupby(['Sex '], as_index=False).count()

Sex   Date  Type  Country  Area  Location  Activity  Injury  Fatal (Y/N)
0    .     1     1        1     1         1         1       1            1
1    F   633   633      630   612       604       582     633          575
2    M  5058  5054     5018  4691      4621      4731    5055         4655
3   M      2     2        2     2         2         2       2            2
4    N     2     2        1     1         1         2       2            2
5  lli     1     1        1     1         1         1       1            1

In [632]:
sharkdf['Sex '] = sharkdf['Sex '].replace(['M '],'M')

In [633]:
sharkdf.groupby(['Sex '], as_index=False).count()

Sex   Date  Type  Country  Area  Location  Activity  Injury  Fatal (Y/N)
0    .     1     1        1     1         1         1       1            1
1    F   633   633      630   612       604       582     633          575
2    M  5060  5056     5020  4693      4623      4733    5057         4657
3    N     2     2        1     1         1         2       2            2
4  lli     1     1        1     1         1         1       1            1

In [634]:
#deleting outliers for organizational sake
sharkdf = sharkdf.drop(sharkdf.loc[sharkdf['Sex '] == '.'].index)
sharkdf = sharkdf.drop(sharkdf.loc[sharkdf['Sex '] == 'lli'].index)
sharkdf = sharkdf.drop(sharkdf.loc[sharkdf['Sex '] == 'N'].index)

In [635]:
sharkdf.groupby(['Sex '], as_index=False).count()

Sex   Date  Type  Country  Area  Location  Activity  Injury  Fatal (Y/N)
0    F   633   633      630   612       604       582     633          575
1    M  5060  5056     5020  4693      4623      4733    5057         4657

In [636]:
sharkdf.groupby(['Type'], as_index=False).count()

Type  Date  Country  Area  Location  Activity  Sex   Injury   
0          Boat   134      134   125       127       113    30     134  \
1       Boating   196      195   180       183       176   101     196   
2       Boatomg     1        1     1         1         1     1       1   
3       Invalid   546      541   499       493       424   472     539   
4      Provoked   570      568   537       525       535   514     570   
5  Questionable     2        2     2         2         2     2       2   
6  Sea Disaster   237      229   171       166       226   173     237   
7    Unprovoked  4537     4506  4263      4199      4237  4396    4536   

   Fatal (Y/N)  
0          133  
1          195  
2            1  
3           22  
4          568  
5            2  
6          233  
7         4531

In [637]:
sharkdf['Type'] = sharkdf['Type'].replace(['Boat', 'Boatomg'],'Boating')

In [638]:
sharkdf.groupby(['Type'], as_index=False).count()

Type  Date  Country  Area  Location  Activity  Sex   Injury   
0       Boating   331      330   306       311       290   132     331  \
1       Invalid   546      541   499       493       424   472     539   
2      Provoked   570      568   537       525       535   514     570   
3  Questionable     2        2     2         2         2     2       2   
4  Sea Disaster   237      229   171       166       226   173     237   
5    Unprovoked  4537     4506  4263      4199      4237  4396    4536   

   Fatal (Y/N)  
0          329  
1           22  
2          568  
3            2  
4          233  
5         4531

In [639]:
sharkdf.groupby(['Activity'], as_index=False).count()

Activity  Date  Type  Country   
0                                                           1     1        1  \
1                                                           1     1        1   
2      a canoe was pursuing a schooner that had forc...     1     1        1   
3                                       "Boat accident"     1     1        1   
4     "Climbing up to ship after repairing the stern...     1     1        1   
...                                                 ...   ...   ...      ...   
1521  ship torpedoed 400 miles off the African coas....     1     1        1   
1522                                         small boat     1     1        1   
1523        wreck of the State Oil Company ship Permina     1     1        1   
1524                                  yachting accident     1     1        1   
1525                              yachtsman in a zodiac     1     1        1   

      Area  Location  Sex   Injury  Fatal (Y/N)  
0        1         1     0       1            1  
1        0         1     1       1            1  
2        0         0     1       1            1  
3        1         1     1       1            1  
4        0         0     1       1            1  
...    ...       ...   ...     ...          ...  
1521     0         0     1       1            1  
1522     1         1     0       1            0  
1523     1         1     0       1            1  
1524     1         1     1       1            1  
1525     1         1     1       1            1  

[1526 rows x 9 columns]

In [640]:
sharkdf.loc[sharkdf['Activity'] == ' ']

Date      Type Country         Area   
5560  Late Jul-1900  Provoked     USA  Connecticut  \

                          Location Activity Sex    
5560  Bridgeport, Fairfield County           NaN  \

                                                 Injury Fatal (Y/N)  
5560  No injury to occupants. They shot shark, then ...           N

In [641]:
sharkdf = sharkdf.drop(sharkdf.loc[sharkdf['Activity'] == ' '].index)

In [642]:
sharkdf.loc[sharkdf['Activity'] == '   ']

Date     Type        Country Area                Location Activity   
2511  4-Jan-93  Boating  CARIBBEAN SEA  NaN  Off Dominican Republic           \

     Sex                                              Injury Fatal (Y/N)  
2511    M  No injury to occupants. Sharks, attracted to o...           N

In [643]:
sharkdf = sharkdf.drop(sharkdf.loc[sharkdf['Activity'] == '   '].index)

In [644]:
sharkdf.groupby(['Activity'], as_index=False).count()

Activity  Date  Type  Country   
0      a canoe was pursuing a schooner that had forc...     1     1        1  \
1                                       "Boat accident"     1     1        1   
2     "Climbing up to ship after repairing the stern...     1     1        1   
3                                  "Crossing the river"     1     1        1   
4     "Flying Tiger" transport plane went down with ...     2     2        2   
...                                                 ...   ...   ...      ...   
1519  ship torpedoed 400 miles off the African coas....     1     1        1   
1520                                         small boat     1     1        1   
1521        wreck of the State Oil Company ship Permina     1     1        1   
1522                                  yachting accident     1     1        1   
1523                              yachtsman in a zodiac     1     1        1   

      Area  Location  Sex   Injury  Fatal (Y/N)  
0        0         0     1       1            1  
1        1         1     1       1            1  
2        0         0     1       1            1  
3        1         1     1       1            1  
4        2         2     2       2            2  
...    ...       ...   ...     ...          ...  
1519     0         0     1       1            1  
1520     1         1     0       1            0  
1521     1         1     0       1            1  
1522     1         1     1       1            1  
1523     1         1     1       1            1  

[1524 rows x 9 columns]

In [645]:
#see if it is worthwhile to use regex to separate into categories
import regex as re
def find_phrase(p):
    return(lambda x: re.findall(r'{}'.format(p), x))


def get_phrase_count(phrase, data, col):
    data_new = data.dropna() #added dropna to ignore null entries
    phrase_results = data_new[col].apply(find_phrase(phrase)) 
    phraselist = list(filter(None, phrase_results))
    phraseCount = len(phraselist)
    return phraseCount

def get_df_record(phrase, data, col):
    data_new = data.dropna() #added dropna to ignore null entries
    phrase_results = data_new[col].dropna().apply(find_phrase(r'{}'.format(phrase))) 
    phrase_bool = pd.Series([bool(x) for x in phrase_results], index = data_new.index)
    df_recs = data_new[phrase_bool]
    return df_recs

In [646]:
get_df_record('yacht', sharkdf, 'Activity')

Date          Type      Country             Area   
3264              2-Jul-74  Sea Disaster          USA         Florida   \
3289                Dec-73    Unprovoked       MEXICO         Guerrero   
3726             30-Nov-63    Unprovoked  NEW ZEALAND     South Island   
4267                  1956    Unprovoked       GREECE     Corfu Island   
4435  Reported 02-Sep-1951    Unprovoked    AUSTRALIA       Queensland   
5773           13-Dec-1884  Sea Disaster    AUSTRALIA  South Australia   
6218               No date    Unprovoked      REUNION       Grand'Anse   

                              Location   
3264                    Gulf of Mexico  \
3289                      Acapulco Bay   
3726  Charteris Bay, Lyttleton Harbour   
4267                           Kérkira   
4435    Fitzroy River near Rockhampton   
5773                      Port Phillip   
6218                        Petite-île   

                                               Activity Sex    
3264  Adrift after the sinking of the motor yacht Pr...    M  \
3289             Swimming alongside yacht Mexico Fiesta    M   
3726      Treading water while alongside capsized yacht    M   
4267                                 Swimming off yacht    F   
4435  Body found on deserted luxury yacht, 38’ Chris...    M   
5773                                  yachting accident    M   
6218                              yachtsman in a zodiac    M   

                                                 Injury Fatal (Y/N)  
3264                                   Arm bitten FATAL           Y  
3289                                              FATAL           Y  
3726                  Hand & lower leg severely injured           N  
4267                                              FATAL           Y  
4435  FATAL, abdominal wounds & right leg severed at...           Y  
5773  Cause of death most likely drowning, remains r...           Y  
6218                                           Survived           N

In [647]:
get_df_record('Boating', sharkdf, 'Activity')

Date        Type  Country          Area   
5102         Late Jul-1928    Provoked    ITALY       Tuscany  \
5434  Reported 08-Jul-1908  Unprovoked  CROATIA  Adriatic Sea   
5775  Reported 28-Aug-1884  Unprovoked      USA    New Jersey   

                    Location Activity Sex    
5102               Viareggio  Boating    M  \
5434                  Medola  Boating    F   
5775  Bayonne, Hudson County  Boating    M   

                                                 Injury Fatal (Y/N)  
5102  No injury, shark approached the boat, he hit i...           N  
5434                                  FATAL, lost a leg           Y  
5775                                        Hand bitten           N

In [648]:
#regex can yield very unhelpful, ambiguous results
get_df_record('ship', sharkdf, 'Activity')

Date          Type               Country   
1189              25-Nov-08  Sea Disaster           PHILIPPINES  \
1321              30-Sep-07  Sea Disaster           PHILIPPINES   
1713              13-Dec-03    Unprovoked           NEW ZEALAND   
4038              19-Dec-59  Sea Disaster           PHILIPPINES   
4114              17-Jan-59    Unprovoked             AUSTRALIA   
4131              12-Dec-58    Unprovoked        AMERICAN SAMOA   
4197              11-May-57    Unprovoked             AUSTRALIA   
4532              14-Dec-48    Unprovoked                  CUBA   
4669              18-Jul-43  Sea Disaster                   USA   
4691              28-Nov-42  Sea Disaster          SOUTH AFRICA   
4701              12-Oct-42  Sea Disaster       SOLOMON ISLANDS   
4749              29-Jan-41    Unprovoked  SOUTH ATLANTIC OCEAN   
4809              29-Aug-38    Unprovoked                 CHINA   
5088               5-Jan-29    Unprovoked                  FIJI   
5173                   1925    Unprovoked                  FIJI   
5381              26-Oct-11    Unprovoked                   USA   
5424               6-Mar-09  Sea Disaster          SOUTH AFRICA   
5432   Reported 28-Aug-1908    Unprovoked                 SPAIN   
5503               2-Nov-04    Unprovoked           PHILIPPINES   
5538  .Reported 22-Feb-1902      Provoked           NEW ZEALAND   
5664            03-Jul-1893  Sea Disaster           PHILIPPINES   
5720               Jan-1889    Unprovoked          SIERRA LEONE   
5725            14-Aug-1888  Sea Disaster                CANADA   
5746            20-Jan-1887  Sea Disaster                BRAZIL   
5847            13-Feb-1878  Sea Disaster          SOUTH AFRICA   
5928               Aug-1867    Unprovoked                MEXICO   
5929            26-Jun-1867  Sea Disaster                  CUBA   
5931                   1867    Unprovoked                  CUBA   
5952            14-Dec-1863    Unprovoked                GUINEA   
5986            11-Mar-1860    Unprovoked               BAHAMAS   
6000            30-Sep-1855  Sea Disaster                   USA   
6021            26-Feb-1852  Sea Disaster          SOUTH AFRICA   
6094            10-Jun-1829    Unprovoked                   USA   
6096            28-Sep-1828    Unprovoked          SIERRA LEONE   
6251            Before 1958    Unprovoked             INDONESIA   
6284            Before 1911    Unprovoked               VIETNAM   
6285            Before 1921    Unprovoked          SOUTH AFRICA   

                                 Area   
1189                  Batanes Provine  \
1321                          Palawan   
1713                     Cook Islands   
4038                          Masbate   
4114                         Tasmania   
4131                   Tutuila Island   
4197                  New South Wales   
4532              Guantanamo Province   
4669                          Florida   
4691                    KwaZulu-Natal   
4701            Makora-Ulawa Province   
4749                 In Convoy OB 274   
4809                      North China   
5088                 Viti Levu Island   
5173                 Viti Levu Island   
5381                   South Carolina   
5424             South Atlantic Ocean   
5432                          Galicia   
5503                     Luzon Island   
5538                     North Island   
5664                     Luzon Island   
5720                     Western Area   
5725                      Nova Scotia   
5746                          Alagoas   
5847            Western Cape Province   
5928                        Veracruz    
5929             Villa Clara Province   
5931  Matanzas Province (north coast)   
5952                   Conakry Region   
5986            New Providence Island   
6000                   North Carolina   
6021            Western Cape Province   
6094                       New Jersey   
6096                     Western Area   
6251                    Riau Province   


In [649]:
#another example
get_df_record('swimming', sharkdf, 'Activity')

Date          Type           Country   
752              15-Aug-12    Unprovoked               USA  \
1669             23-May-04    Unprovoked               USA   
2090              6-Nov-99    Unprovoked            BRAZIL   
2694              2-Feb-89    Unprovoked             ITALY   
3001             13-Dec-81    Unprovoked               USA   
3467             10-Jul-68    Unprovoked           BAHAMAS   
3535              6-Feb-67  Sea Disaster            MEXICO   
3652              9-Dec-64    Unprovoked  PAPUA NEW GUINEA   
3660              4-Oct-64    Unprovoked         AUSTRALIA   
3937              1-Jan-61    Unprovoked          COLUMBIA   
4168             27-Jan-58    Unprovoked         AUSTRALIA   
4316             29-Dec-54    Unprovoked    AMERICAN SAMOA   
4668                Aug-43  Sea Disaster             ITALY   
4789  Reported 25-Oct-1939    Unprovoked         AUSTRALIA   
4851             15-May-37    Unprovoked         AUSTRALIA   
4878             26-Jun-36    Unprovoked         AUSTRALIA   
5183             25-Apr-24    Unprovoked         AUSTRALIA   
5192              9-Jan-24    Unprovoked         AUSTRALIA   
5255              8-Mar-20    Unprovoked         AUSTRALIA   
5266             12-Sep-19    Unprovoked        COSTA RICA   
5384             20-Sep-11    Unprovoked               USA   
5601                  1898    Unprovoked              CUBA   
5670           14-Apr-1893    Unprovoked         AUSTRALIA   
6272          World War II  Sea Disaster  PAPUA NEW GUINEA   
6285           Before 1921    Unprovoked      SOUTH AFRICA   

                               Area   
752                         Alabama  \
1669                        Florida   
2090                 Rio de Janeiro   
2694                 Tyrrhenian Sea   
3001                         Hawaii   
3467             Great Exuma Island   
3535                       Veracruz   
3652  Bougainville (North Solomons)   
3660              Western Australia   
3937                  Roncador Bank   
4168                New South Wales   
4316                 Tutuila Island   
4668      40 miles south of Naples    
4789                     Queensland   
4851                     Queensland   
4878                  Torres Strait   
5183                New South Wales   
5192                New South Wales   
5255                     Queensland   
5266                   Turtle Bogue   
5384                        Florida   
5601              Camaguey Province   
5670                     Queensland   
6272                Madang Province   
6285                  KwaZulu-Natal   

                                               Location   
752                         Gulf Shores, Baldwin County  \
1669                   St. Augustine, St. John's County   
2090                      Rio de Janeiro, Guanabara Bay   
2694          Golfo di Baratti, near Piombino (Tuscany)   
3001                 Nimitz Beach, Barbers Point, O'ahu   
3467                Children's Bay Cay near Rollerville   
3535                                          Cabo Rojo   
3652                      Off Sing Village, Buka Island   
3660  Avalon Beach, near Mandurah, 50 miles south of...   
3937         Roncador Bank, 135 nm north of San Andres    
4168  5 miles south of Sussex Inlet at mouth of Bera...   
4316              Near tuna cannery in Pago Pago Harbor   
4668                                     Tyrrhenian Sea   
4789          Near Restoration Rock off Portland Roads    
4851                             Ross River, Townsville   
4878                                      Nepean Island   
5183                                              Kiama   
5192    Near Asbestos Works, Camellia, Parramatta River   
5255  Between Bay Rock & Magnetic Island, Cleveland Bay   
5266                Where Colorado River enters the sea   
5384                     Pensacola Bay, Escambia County   
5601                                    Nuevitas Harbor   
5670                                 

In [650]:
#deleting activity
del sharkdf['Activity']

In [651]:
get_df_record('FATAL', sharkdf, 'Injury')

Date        Type             Country   
6               3-Jun-18  Unprovoked              BRAZIL  \
58             30-Nov-17  Unprovoked          COSTA RICA   
64              4-Nov-17  Unprovoked                CUBA   
91              3-Sep-17  Unprovoked        SOUTH AFRICA   
144             5-May-17  Unprovoked              MEXICO   
...                  ...         ...                 ...   
6290  Before 19-Jul-1913  Unprovoked        SOUTH AFRICA   
6296         Before 1906  Unprovoked           AUSTRALIA   
6297         Before 1903  Unprovoked           AUSTRALIA   
6299           1900-1905  Unprovoked                 USA   
6301           1845-1853  Unprovoked  CEYLON (SRI LANKA)   

                       Area                             Location Sex    
6                Pernambuco                Piedade Beach, Recife    M  \
58             Cocos Island                            Manuelita    F   
64         Holquin Province                   Guardalavaca Beach    M   
91    Western Cape Province                          Dyer Island    M   
144     Baja California Sur            Los Arbolitos, Cabo Pulmo    M   
...                     ...                                  ...  ...   
6290          KwaZulu-Natal                               Durban    M   
6296        New South Wales                                         M   
6297      Western Australia                          Roebuck Bay    M   
6299         North Carolina                       Ocracoke Inlet    M   
6301       Eastern Province  Below the English fort, Trincomalee    M   

                                                 Injury Fatal (Y/N)  
6                                                 FATAL           Y  
58                                                FATAL           Y  
64                                                FATAL           Y  
91                                                FATAL           Y  
144                                               FATAL           Y  
...                                                 ...         ...  
6290                     FATAL, leg stripped of flesh             Y  
6296                                              FATAL           Y  
6297                                              FATAL           Y  
6299                                              FATAL           Y  
6301  FATAL. "Shark bit him in half, carrying away t...           Y  

[988 rows x 8 columns]

In [652]:
sharkdf['Injury'].value_counts()['FATAL'] #redundant information in the injury column along with inconsistent patterns of descriptions.

802

In [653]:
#deleting Injury Column
del sharkdf['Injury']

In [654]:
sharkdf.head()

Date        Type    Country             Area   
0  25-Jun-18     Boating        USA       California  \
1  18-Jun-18  Unprovoked        USA          Georgia   
2   9-Jun-18     Invalid        USA           Hawaii   
3   8-Jun-18  Unprovoked  AUSTRALIA  New South Wales   
4   4-Jun-18    Provoked     MEXICO           Colima   

                         Location Sex  Fatal (Y/N)  
0     Oceanside, San Diego County    F           N  
1  St. Simon Island, Glynn County    F           N  
2                    Habush, Oahu    M           N  
3              Arrawarra Headland    M           N  
4                        La Ticla    M           N

In [655]:
get_df_record('No date', sharkdf, 'Date')

Date        Type   
6217                                            No date  Unprovoked  \
6218                                            No date  Unprovoked   
6220                           No date, Before May-1996  Unprovoked   
6221                           No date, Before Mar-1995  Unprovoked   
6223                           No date, Before Aug-1989  Unprovoked   
6224                           No date, Before Aug-1987    Provoked   
6225                               No date, Before 1987  Unprovoked   
6226                              No date, Before  1975    Provoked   
6227                               No date, Before 1975     Invalid   
6229                         No date, Before 3-Jan-1967  Unprovoked   
6230                               No date, Before 1963  Unprovoked   
6231                         No date, Before 8-May-1965  Unprovoked   
6233                               No date, Before 1902  Unprovoked   
6236         No date, After August 1926 and before 1936  Unprovoked   
6275                                            No date  Unprovoked   
6279                                            No date  Unprovoked   
6281                                            No date  Unprovoked   
6289  No date (3 days after preceding incident) & pr...  Unprovoked   

               Country                   Area   
6217               USA                Florida  \
6218           REUNION             Grand'Anse   
6220             KOREA            South Korea   
6221  FRENCH POLYNESIA               Tuamotus   
6223           VANUATU       Malampa Province   
6224           VANUATU       Malampa Province   
6225              IRAN     Khuzestan Province   
6226               USA                Florida   
6227  PAPUA NEW GUINEA    Milne Bay  Province   
6229      SOUTH AFRICA  Eastern Cape Province   
6230               USA             California   
6231            GREECE        Island of Volos   
6233               USA                Florida   
6236         AUSTRALIA      Western Australia   
6275            GREECE     Dodecanese Islands   
6279      SOUTH AFRICA  Western Cape Province   
6281      SOUTH AFRICA          KwaZulu-Natal   
6289      SOUTH AFRICA          KwaZulu-Natal   

                                              Location Sex  Fatal (Y/N)  
6217                           Key West, Monroe County    M           N  
6218                                        Petite-île    M           N  
6220                                      Cheju Island    F           Y  
6221                                          Rangiroa    M           N  
6223                                          Malakula    F           Y  
6224                                   Hokai, Malakula    M           N  
6225                         Ahvaz, on the Karun River    M           N  
6226                  Riviera Beach, Palm Beach County    M           N  
6227   D'Entrecasteaux islands, 20 miles off the coast    M           Y  
6229                             Keiskamma River mouth    M           Y  
6230                         LaJolla, San Diego County    M           N  
6231                                     Eastern shore    F           Y  
6233      Mosquito Inlet (Ponce Inlet), Volusia County    M           Y  
6236                                     Cossack Creek    M           N  
6275                                       Symi Island    M           N  
6279                                          Arniston    F           N  
6281                                           Durban     M           Y  
6289                                            Durban    M           Y

In [656]:
dateDeldf = get_df_record('No date', sharkdf, 'Date')
sharkdf = sharkdf.drop(dateDeldf.index)

In [657]:
dateDeldf_bf = get_df_record('Before', sharkdf, 'Date')
sharkdf = sharkdf.drop(dateDeldf_bf.index)

In [658]:
get_df_record('&', sharkdf, 'Date')

Date          Type      Country   
4631  Between 18 & 22-Dec 1944  Sea Disaster  PHILIPPINES  \

                         Area   
4631  300 miles east of Luzon  \

                                               Location Sex  Fatal (Y/N)  
4631  USS Hull DD-350 was one of 3 destroyers that c...    M           N

In [659]:
sharkdf = sharkdf.replace(['Between 18 & 22-Dec 1944'],'20-Dec-44')

In [660]:
get_df_record('20-Dec-44', sharkdf, 'Date')

Date          Type      Country                     Area   
4631  20-Dec-44  Sea Disaster  PHILIPPINES  300 miles east of Luzon  \

                                               Location Sex  Fatal (Y/N)  
4631  USS Hull DD-350 was one of 3 destroyers that c...    M           N

In [661]:
sharkdf.groupby(['Date'], as_index=False).count() #more ambiguous dates and inconsistencies 

Date  Type  Country  Area  Location  Sex   Fatal (Y/N)
0      \n1951.12.15.R     1        1     1         1     1            0
1         10-Jan-2009     1        1     1         1     0            1
2         15-Jun-1937     1        1     1         1     1            1
3         16-Jan-1970     1        1     1         1     1            1
4         22-Jul-2013     1        1     1         1     1            1
...               ...   ...      ...   ...       ...   ...          ...
5290      Summer-2008     1        1     1         1     1            1
5291      Winter 1942     1        1     1         1     1            1
5292      Winter 1969     1        1     1         0     0            1
5293    Woirld War II     1        1     0         0     1            1
5294     World War II     2        2     2         2     2            2

[5295 rows x 7 columns]

In [662]:
#deleting Date and removing time series element completely
del sharkdf['Date']

In [663]:
sharkdf.head()

Type    Country             Area                        Location   
0     Boating        USA       California     Oceanside, San Diego County  \
1  Unprovoked        USA          Georgia  St. Simon Island, Glynn County   
2     Invalid        USA           Hawaii                    Habush, Oahu   
3  Unprovoked  AUSTRALIA  New South Wales              Arrawarra Headland   
4    Provoked     MEXICO           Colima                        La Ticla   

  Sex  Fatal (Y/N)  
0    F           N  
1    F           N  
2    M           N  
3    M           N  
4    M           N

In [664]:
sharkdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6166 entries, 0 to 6301
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Type         6162 non-null   object
 1   Country      6119 non-null   object
 2   Area         5720 non-null   object
 3   Location     5638 non-null   object
 4   Sex          5633 non-null   object
 5   Fatal (Y/N)  5627 non-null   object
dtypes: object(6)
memory usage: 337.2+ KB


In [673]:
sharkdf['Fatal (Y/N)'] = sharkdf['Fatal (Y/N)'].map({'N': 0,'Y': 1})

In [674]:
sharkdf['Fatal (Y/N)'] = sharkdf['Fatal (Y/N)'].astype('bool')*1

In [675]:
sharkdf = sharkdf.rename(columns={'Fatal (Y/N)': 'Fatal'})

In [676]:
sharkdf = sharkdf.rename(columns={'Sex ': 'Sex'})

In [677]:
sharkdf.head()

Type    Country             Area                        Location Sex   
0     Boating        USA       California     Oceanside, San Diego County   F  \
1  Unprovoked        USA          Georgia  St. Simon Island, Glynn County   F   
2     Invalid        USA           Hawaii                    Habush, Oahu   M   
3  Unprovoked  AUSTRALIA  New South Wales              Arrawarra Headland   M   
4    Provoked     MEXICO           Colima                        La Ticla   M   

   Fatal  
0      0  
1      0  
2      0  
3      0  
4      0

In [678]:
sharkdf.head(35)

Type        Country                   Area   
0        Boating            USA             California  \
1     Unprovoked            USA                Georgia   
2        Invalid            USA                 Hawaii   
3     Unprovoked      AUSTRALIA        New South Wales   
4       Provoked         MEXICO                 Colima   
5     Unprovoked      AUSTRALIA        New South Wales   
6     Unprovoked         BRAZIL             Pernambuco   
7     Unprovoked            USA                Florida   
8     Unprovoked            USA                Florida   
9     Unprovoked            USA                Florida   
10      Provoked      AUSTRALIA             Queensland   
11    Unprovoked            USA         South Carolina   
12    Unprovoked            USA         South Carolina   
13       Invalid        ENGLAND               Cornwall   
14      Provoked      AUSTRALIA      Westerm Australia   
15    Unprovoked   SOUTH AFRICA  Eastern Cape Province   
16  Questionable      AUSTRALIA        New South Wales   
17    Unprovoked       THAILAND                Hua Hin   
18    Unprovoked     COSTA RICA           Cocos Island   
19    Unprovoked      AUSTRALIA      Western Australia   
20  Questionable      AUSTRALIA        New South Wales   
21    Unprovoked      AUSTRALIA      Western Australia   
22    Unprovoked      AUSTRALIA      Western Australia   
23    Unprovoked       MALDIVES      Alifu Alifu Atoll   
24    Unprovoked   SOUTH AFRICA  Western Cape Province   
25    Unprovoked            USA                 Hawaii   
26    Unprovoked       THAILAND                Hua Hin   
27    Unprovoked         BRAZIL             Pernambuco   
28    Unprovoked      AUSTRALIA      Western Australia   
29    Unprovoked      AUSTRALIA      Western Australia   
30    Unprovoked        BAHAMAS         New Providence   
31       Invalid         BRAZIL                Alagoas   
32    Unprovoked  NEW CALEDONIA                    NaN   
33    Unprovoked        BAHAMAS                    NaN   
34    Unprovoked   SOUTH AFRICA  Eastern Cape Province   

                                             Location  Sex  Fatal  
0                         Oceanside, San Diego County    F      0  
1                      St. Simon Island, Glynn County    F      0  
2                                        Habush, Oahu    M      0  
3                                  Arrawarra Headland    M      0  
4                                            La Ticla    M      0  
5                                  Flat Rock, Ballina    M      0  
6                               Piedade Beach, Recife    M      1  
7   Lighhouse Point Park, Ponce Inlet, Volusia County    M      0  
8                        Cocoa Beach, Brevard  County    M      0  
9                       Daytona Beach, Volusia County    M      0  
10                                    Cairns Aquarium    M      0  
11                  Isle of Palms,  Charleston County    M      0  
12                Hilton Head Island, Beaufort County    M      0  
13                                     Off Land's End    M      0  
14                                         Dugong Bay    F      0  
15                          Nahoon Beach, East London    M      0  
16                             Sharpes Beach, Ballina    M      0  
17                                      Sai Noi Beach    M      0  
18                                          Manuelita    M      0  
19                                       Gearys Beach    M      0  
20                                        Lennox Head    M      0  
21                            Surfers Point, Prevelly    M      0  
22                             South Point, Gracetown    M      0  
23                                          Madoogali    M      0  
24                     Robberg Beach, Plettenberg Bay    M      0  
25             Shipwreck’s Beach, Keoneloa Bay, Kauai    M      0  
26                                      Sai Noi Beach    M      0  
27                 

In [679]:
sharkdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6166 entries, 0 to 6301
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      6162 non-null   object
 1   Country   6119 non-null   object
 2   Area      5720 non-null   object
 3   Location  5638 non-null   object
 4   Sex       5633 non-null   object
 5   Fatal     6166 non-null   int32 
dtypes: int32(1), object(5)
memory usage: 313.1+ KB


Train and Test Logistic Regression Model

In [680]:
from sklearn.model_selection import train_test_split
train_data , test_data = train_test_split(sharkdf, test_size=0.2, random_state=38)

In [681]:
import statsmodels.api as sm
from statsmodels.formula.api import logit 

In [682]:
train_data

Type           Country                   Area   
4443    Unprovoked  FRENCH POLYNESIA               Tuamotus  \
3143       Invalid               USA                Florida   
5630  Sea Disaster              CUBA        Havana Province   
2865    Unprovoked     NEW CALEDONIA         South Province   
5084    Unprovoked         AUSTRALIA             Queensland   
...            ...               ...                    ...   
5435    Unprovoked             EGYPT           Gulf of Suez   
4457    Unprovoked         AUSTRALIA             Queensland   
5648    Unprovoked         AUSTRALIA        New South Wales   
4189    Unprovoked   SOLOMON ISLANDS                Honiara   
3954       Boating      SOUTH AFRICA  Western Cape Province   

                                               Location  Sex  Fatal  
4443  Outer edge of Hikueru, one of the Central Tuam...    M      0  
3143                    St. Petersburg, Pinellas County    M      1  
5630                                      Havana Harbor    M      1  
2865                                      Amedee Island  NaN      0  
5084              Alma Bay, Magnetic Island, Townsville    M      1  
...                                                 ...  ...    ...  
5435                                     Off Ras Gharib    M      1  
4457  Palm Beach North, 5 miles north of Burleigh He...    M      1  
5648                                         Newcastle     M      0  
4189                                            Honiara    M      1  
3954                      Jacobs Bay Reef, Saldanha Bay  NaN      0  

[4932 rows x 6 columns]

Use of all variables prompted 'Maximum number of iterations has been exceeded' error. Geographical data had to be excluded in order to avoid the error. The data set was already shown to have several issues with structure and consistency, making it difficult to use in a comprehensive way.

In [690]:
formula = ('Fatal ~ Type + Sex')

model = logit(formula = formula, data = train_data).fit()

Optimization terminated successfully.
         Current function value: 0.486811
         Iterations 7


In [689]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  Fatal   No. Observations:                 4515
Model:                          Logit   Df Residuals:                     4509
Method:                           MLE   Df Model:                            5
Date:                Sun, 07 Jan 2024   Pseudo R-squ.:                  0.2049
Time:                        11:13:38   Log-Likelihood:                -2198.0
converged:                       True   LL-Null:                       -2764.4
Covariance Type:            nonrobust   LLR p-value:                1.027e-242
========================================================================================
                           coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept               -3.0289      0.411     -7.363      0.000      -3.835      -2.223
Type[T.Invalid]          5.8806      0.478     12.297      0.000       4.943       6.818
Type[T.Provoked]        -0.6045      0.464     -1.302      0.193      -1.514       0.305
Type[T.Sea Disaster]     3.2505      0.431      7.538      0.000       2.405       4.096
Type[T.Unprovoked]       1.4850      0.395      3.764      0.000       0.712       2.258
Sex[T.M]                 0.5048      0.129      3.910      0.000       0.252       0.758
========================================================================================
"""

In [691]:
#logistic regression interpretation tutorial
#https://onezero.blog/modelling-binary-logistic-regression-using-python-research-oriented-modelling-and-interpretation/

import numpy as np
print(np.exp(model.params))

Intercept                 0.048371
Type[T.Invalid]         358.025309
Type[T.Provoked]          0.546373
Type[T.Sea Disaster]     25.804417
Type[T.Unprovoked]        4.415060
Sex[T.M]                  1.656658
dtype: float64


This shows how much the lack of data skews the results of the logistic regression. If we were to interpret this without context, the invalid category indicates one unit increase in the invalid category of attack type increases the odds of a fatal attack by 358 times.

This is a frustrating example of where the sample size is not the issue for modeling correctly, but the lack of complete and consistent data.
Near infinite categories of geographical data instead of consolidation into succinct groups resulted in the removal from the model. Hence defeating the original objective of determining fatal attacks by location, area, or country, at the minimum, along with type and sex.